saleor/order/actions.py

```diff
@@ -388,3 +399,144 @@ def create_fulfillments(
+            fulfillment_lines_to_refund, fulfillment_lines_quantities_to_refund
+        ):
+            refunded_line = _get_fulfillment_line_if_exists(
+                already_refunded_lines,
+                fulfillment_line.order_line_id,
+                fulfillment_line.stock_id,
+            )
+            fulfillemt_line_existed = True
+            if not refunded_line:
+                fulfillemt_line_existed = False
+                refunded_line = FulfillmentLine(
+                    fulfillment=refunded_fulfillment,
+                    order_line_id=fulfillment_line.order_line_id,
+                    stock_id=fulfillment_line.stock_id,
+                    quantity=0,
+                )
+            fulfilled_to_refund = min(fulfillment_line.quantity, quantity_to_refund)
+            quantity_to_refund -= fulfilled_to_refund
+            refunded_line.quantity += fulfilled_to_refund
+            fulfillment_line.quantity -= fulfilled_to_refund
+            order_line: OrderLine = order_lines_with_fulfillment.get(  # type: ignore
+                fulfillment_line.order_line_id
+            )
+            refund_amount += order_line.unit_price_gross_amount * fulfilled_to_refund
+            if fulfillment_line.quantity == 0:
+                empty_fulfillment_lines_to_delete.append(fulfillment_line)
+            else:
+                fulfillment_lines_to_update.append(fulfillment_line)
+            if refunded_line.quantity > 0 and not fulfillemt_line_existed:
+                fulfillment_lines_to_create.append(refunded_line)
+            elif fulfillemt_line_existed:
+                fulfillment_lines_to_update.append(refunded_line)
+
+            data_from_all_refunded_lines = all_refunded_lines.get(order_line.id)
+            if data_from_all_refunded_lines:
+                quantity, line = data_from_all_refunded_lines
+                quantity += fulfilled_to_refund
+                all_refunded_lines[order_line.id] = (quantity, line)
+            else:
+                all_refunded_lines[order_line.id] = (
+                    fulfilled_to_refund,
+                    order_line,
+                )  # type: ignore
+
+        OrderLine.objects.bulk_update(order_lines_to_update, ["quantity_fulfilled"])
+        FulfillmentLine.objects.bulk_update(fulfillment_lines_to_update, ["quantity"])
+        FulfillmentLine.objects.bulk_create(fulfillment_lines_to_create)
+        FulfillmentLine.objects.filter(
+            id__in=[f.id for f in empty_fulfillment_lines_to_delete]
+        ).delete()
+        Fulfillment.objects.filter(order=order, lines=None).delete()
+        for line, quantity in lines_to_deallocate_stock:
+            deallocate_stock(line, quantity)
+    if amount is None:
+        amount = refund_amount
+    if refund_shipping_costs:
+        amount += order.shipping_price_gross_amount
+    if amount:
+        amount = min(payment.captured_amount, amount)
+        gateway.refund(payment, amount)
```

Comment:
maybe there should there be a check here like `if payment.can_refund()`?

Reply:
It is in `saleor/graphql/order/mutations/fulfillments.py:411-412
```
payment = order.get_last_payment()
if not payment or not payment.can_refund():
```
Do you suggest to put there an additional checking condition?